src/psyclone/psyir/transformations/loop_swap_trans.py

```diff
@@ -0,0 +1,159 @@
+        :type options: dict of string:values or None
+
+        :raises TransformationError: if the supplied node does not\
+                                     allow a loop swap to be done.
+
+        '''
+        super().validate(node_outer, options=options)
+
+        if not node_outer.loop_body or not node_outer.loop_body.children:
+            raise TransformationError("Error in LoopSwap transformation. "
+                                      "Supplied node '{0}' must be the outer "
+                                      "loop of a loop nest and must have one "
+                                      "inner loop, but this node does not "
+                                      "have any statements inside."
+                                      .format(node_outer))
+
+        node_inner = node_outer.loop_body[0]
+
+        # Check that the body of the outer loop is itself a Loop
+        try:
+            super().validate(node_inner, options=options)
+        except TransformationError as err:
+            raise TransformationError("Error in LoopSwap transformation. "
+                                      "Supplied node '{0}' must be the outer "
+                                      "loop of a loop nest but the first "
+                                      "inner statement is not a valid loop:\n"
+                                      "{1}.".format(node_outer,
+                                                    str(err.value))) from err
+
+        if len(node_outer.loop_body.children) > 1:
+            raise TransformationError(
+                "Error in LoopSwap transformation. Supplied node '{0}' must"
+                " be the outer loop of a loop nest and must have exactly one "
+                "inner loop, but this node has {1} inner statements, the "
+                "first two being '{2}' and '{3}'"
+                "".format(node_outer, len(node_outer.loop_body.children),
+                          node_outer.loop_body[0], node_outer.loop_body[1]))
+
+    def apply(self, outer, options=None):
+        # pylint: disable=arguments-differ
+        '''The argument :py:obj:`outer` must be a loop which has exactly
+        one inner loop. This transform then swaps the outer and inner loop.
+
+        :param outer: the node representing the outer loop.
+        :type outer: :py:class:`psyclone.psyir.nodes.Loop`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        :raises TransformationError: if the supplied node does not \
+                                     allow a loop swap to be done.
+
+        '''
+        self.validate(outer, options=options)
+
+        inner = outer.loop_body[0]
+
+        # Detach the inner code
+        inner_loop_body = inner.loop_body.detach()
+
+        # Swap the loops
+        outer.replace_with(inner.detach())
+        inner.addchild(Schedule())
```

Comment:
Could this be `inner.addchild(outer.loop_body.detach())`?

outer.loop_body should be an empty Schedule after the previous inner.detach() but it could have an attached symbol_table with a defined symbol (not form Fortran code since there are only subroutine scope symbols but we may have added it manually).  It could also have a node annotation (although there are no schedule annotations for now). So I think its safer to re-use the existing Schedule than try to create a new one.

Also can you create a test where you manually add a symbol to each loop_body symbol table and check that this still exist after the transformation.

Reply:
It doesn't work out of the box (one of the loop is incomplete, to some validation is triggered).

My main question is: should I also swap the schedules, or leave them? If I swap them, I am not sure if we might not get different behaviour (if both loops declare the same variable?? But maybe this would then be detected when we have dependency analysis)? Let's discuss this in the Telco today.

Reply:
Implemented as discussed in the telco:
1. Now the schedules remain in place, no new schedule is created.
2. A TransformationError is raised if one of the two schedules involved contains a non-empty symbol table (every Schedule apparently has an empty symbol table as far as I can tell).

For the test I did not add a symbol, I tested that each old schedule ``is`` the same as new_schedule, i.e. same instance.